mypipe-api/src/main/scala/mypipe/mysql/AbstractMySQLBinaryLogConsumer.scala

```diff
@@ -130,41 +144,41 @@ abstract class AbstractMySQLBinaryLogConsumer
     ThreadLocalRandom.current().nextLong(lower, upper + 1)
   }
 
-  override protected def onStart(): Future[Boolean] = {
-    @volatile var connected = false
-    client.setServerId(nextServerId())
-    client.registerEventListener(new EventListener() {
-      override def onEvent(event: MEvent) = handleEvent(event)
-    })
+  /**
+   * Register a unique instance of this with the client to be notified on connection
+   */
+  private class ConnectionNotifyingLifecycleListener() extends LifecycleListener {
+    val onConnected: Promise[Unit] = Promise()
 
-    client.registerLifecycleListener(new LifecycleListener {
-      override def onDisconnect(client: BinaryLogClient) = handleDisconnect()
+    override def onConnect(client: BinaryLogClient): Unit = if (!onConnected.isCompleted) onConnected.success(())
+    override def onCommunicationFailure(client: BinaryLogClient, ex: Exception): Unit = {}
+    override def onEventDeserializationFailure(client: BinaryLogClient, ex: Exception): Unit = {}
+    override def onDisconnect(client: BinaryLogClient): Unit = {}
 
-      override def onConnect(client: BinaryLogClient) = {
-        log.info(s"connected = true")
-        connected = true
-        handleConnect()
-      }
+    def isConnected: Boolean = onConnected.future.isCompleted
+  }
 
-      override def onEventDeserializationFailure(client: BinaryLogClient, ex: Exception) {}
+  override protected def onStart(): Future[Boolean] = {
+    client.setServerId(nextServerId())
 
-      override def onCommunicationFailure(client: BinaryLogClient, ex: Exception) {}
-    })
+    // Because we need to attach a lifecycle listener each time onStart is called
+    val connectionNotifyingLifecycleListener = new ConnectionNotifyingLifecycleListener()
+    client.registerLifecycleListener(connectionNotifyingLifecycleListener)
+    val isConnected = () => connectionNotifyingLifecycleListener.isConnected
 
     log.info(s"Connecting client to $client.get:$hostname:$port")
-
     // `client.connect()` blocks, so use a Thread instead of a Future
     val clientThread = new Thread(new Runnable {
       def run() {
         client.connect()
       }
     })
-    clientThread.start
+    clientThread.start()
 
     val startTime = Platform.currentTime
-    while (Platform.currentTime - startTime < connectWaitTime && !connected) Thread.sleep(10)
+    while (Platform.currentTime - startTime < connectWaitTime && !isConnected()) Thread.sleep(10)
 
-    Future.successful(connected)
+    Future.successful(isConnected())
   }
 
   override protected def onStop(): Unit = client.disconnect()
```

Comment:
Hmm, do we have ordering guarantees? In other words, is it possible for this to happen:
```
- We call onStop, ie. disconnect client
- We call onStart, ie. setup new client lifecycle listener
- Client updates listeners with connect messages
- Client updates listeners with disconnect message
```

I'm guessing this is likely not possible though. Also if client.disconnect is blocking() this should probably be fine.

Reply:
I had the same thought, though If that was the case, I believe the current code would be vulnerable to this potential race condition as well. Re-writing the retry logic entirely (potentially one level up), is probably the better option medium term.